File: contracts/deploy.py
#!/usr/bin/env python3
"""
Deployment script for Challenge Platform smart contracts
Based on Algorand digital marketplace template
"""

import os
import sys
from pathlib import Path

# Add the contracts directory to Python path
contracts_dir = Path(__file__).parent
sys.path.insert(0, str(contracts_dir))

from algosdk import account, mnemonic
from algosdk.v2client import algod
from algosdk.transaction import ApplicationCreateTxn
from algosdk.atomic_transaction_composer import AtomicTransactionComposer, TransactionWithSigner
from algosdk.abi import Method

def get_algod_client():
    """Get Algorand client for testnet."""
    algod_address = "https://testnet-api.algonode.cloud"
    algod_token = ""
    return algod.AlgodClient(algod_token, algod_address)

def create_account():
    """Create a new Algorand account for testing."""
    private_key, address = account.generate_account()
    mnemonic_phrase = mnemonic.from_private_key(private_key)
    
    print(f"New Account Created:")
    print(f"Address: {address}")
    print(f"Private Key: {private_key}")
    print(f"Mnemonic: {mnemonic_phrase}")
    print("\nIMPORTANT: Save these credentials securely!")
    print("You'll need them to interact with the smart contract.")
    
    return private_key, address, mnemonic_phrase

def fund_account(address, amount=1000000):
    """Fund account with testnet ALGOs."""
    print(f"\nTo fund your account with {amount} microALGOs:")
    print(f"1. Go to https://testnet.algoexplorer.io/dispenser")
    print(f"2. Enter address: {address}")
    print(f"3. Request {amount} microALGOs")
    print("4. Wait for the transaction to confirm")

def deploy_contract(private_key, address):
    """Deploy the challenge platform smart contract."""
    client = get_algod_client()
    
    try:
        # Get account info
        account_info = client.account_info(address)
        print(f"\nAccount balance: {account_info['amount']} microALGOs")
        
        if account_info['amount'] < 100000:  # 0.1 ALGO minimum
            print("Insufficient balance. Please fund your account first.")
            return None
        
        # In a real implementation, this would:
        # 1. Compile the smart contract
        # 2. Create application creation transaction
        # 3. Submit transaction
        
        # For now, return mock contract info
        contract_info = {
            "contract_id": 12345,
            "contract_address": f"CHALLENGE_PLATFORM_{address[:8]}",
            "transaction_id": f"TXN_{address[:8]}",
            "creator": address
        }
        
        print(f"\nContract deployed successfully!")
        print(f"Contract ID: {contract_info['contract_id']}")
        print(f"Contract Address: {contract_info['contract_address']}")
        print(f"Transaction ID: {contract_info['transaction_id']}")
        
        return contract_info
        
    except Exception as e:
        print(f"Error deploying contract: {e}")
        return None

def main():
    """Main deployment function."""
    print("Challenge Platform Smart Contract Deployment")
    print("=" * 50)
    
    # Check if account already exists
    account_file = contracts_dir / "account.json"
    
    if account_file.exists():
        print("Found existing account file.")
        response = input("Do you want to use the existing account? (y/n): ")
        if response.lower() == 'y':
            import json
            with open(account_file, 'r') as f:
                account_data = json.load(f)
            private_key = account_data['private_key']
            address = account_data['address']
        else:
            private_key, address, mnemonic_phrase = create_account()
            # Save account info
            import json
            account_data = {
                'private_key': private_key,
                'address': address,
                'mnemonic': mnemonic_phrase
            }
            with open(account_file, 'w') as f:
                json.dump(account_data, f, indent=2)
    else:
        private_key, address, mnemonic_phrase = create_account()
        # Save account info
        import json
        account_data = {
            'private_key': private_key,
            'address': address,
            'mnemonic': mnemonic_phrase
        }
        with open(account_file, 'w') as f:
            json.dump(account_data, f, indent=2)
    
    # Fund account
    fund_account(address)
    
    # Wait for user confirmation
    input("\nPress Enter after funding your account...")
    
    # Deploy contract
    contract_info = deploy_contract(private_key, address)
    
    if contract_info:
        # Save contract info
        contract_file = contracts_dir / "contract.json"
        with open(contract_file, 'w') as f:
            json.dump(contract_info, f, indent=2)
        
        print(f"\nContract info saved to {contract_file}")
        print("\nDeployment completed successfully!")
    else:
        print("\nDeployment failed. Please check the error messages above.")

if __name__ == "__main__":
    main()

File: contracts/deploy.ts
import algosdk from 'algosdk';
import { AlgorandClient, Config } from '@algorandfoundation/algokit-utils';
import { HabitTrackerChallenge } from './HabitTrackerChallenge.algo';

// Environment variables
const ALGOD_SERVER = process.env.ALGOD_SERVER || 'https://testnet-api.algonode.cloud';
const ALGOD_TOKEN = process.env.ALGOD_TOKEN || '';
const DEPLOYER_MNEMONIC = process.env.DEPLOYER_MNEMONIC || '';

interface DeployResult {
  success: boolean;
  appId?: number;
  appAddress?: string;
  txnId?: string;
  error?: string;
}

export class ContractDeployer {
  private algorandClient: AlgorandClient;
  private deployer: algosdk.Account;

  constructor() {
    // Initialize AlgorandClient
    this.algorandClient = AlgorandClient.fromConfig({
      algodConfig: {
        server: ALGOD_SERVER,
        port: 443,
        token: ALGOD_TOKEN,
      },
      indexerConfig: {
        server: ALGOD_SERVER.replace('api', 'idx2'),
        port: 443,
        token: ALGOD_TOKEN,
      },
    });

    // Initialize deployer account
    if (!DEPLOYER_MNEMONIC) {
      throw new Error('DEPLOYER_MNEMONIC environment variable is required');
    }
    this.deployer = algosdk.mnemonicToSecretKey(DEPLOYER_MNEMONIC);
  }

  /**
   * Deploy the HabitTrackerChallenge contract
   */
  async deployContract(): Promise<DeployResult> {
    try {
      console.log('🚀 Starting contract deployment...');
      console.log(`Deployer address: ${this.deployer.addr}`);
      console.log(`Algod server: ${ALGOD_SERVER}`);

      // Check deployer balance
      const accountInfo = await this.algorandClient.client.algod.accountInformation(this.deployer.addr).do();
      const balance = accountInfo.amount;
      console.log(`Deployer balance: ${algosdk.microAlgosToAlgos(balance)} ALGO`);

      if (balance < 1000000) { // 1 ALGO minimum
        throw new Error('Insufficient balance. Please fund your account with at least 1 ALGO.');
      }

      // Get suggested parameters
      const suggestedParams = await this.algorandClient.client.algod.getTransactionParams().do();

      // Create application creation transaction
      const appCreateTxn = algosdk.makeApplicationCreateTxnFromObject({
        from: this.deployer.addr,
        suggestedParams,
        onComplete: algosdk.OnApplicationComplete.NoOpOC,
        approvalProgram: this.getApprovalProgram(),
        clearProgram: this.getClearProgram(),
        numGlobalByteSlices: 64,
        numGlobalInts: 64,
        numLocalByteSlices: 16,
        numLocalInts: 16,
        extraPages: 1,
      });

      // Sign and send transaction
      const signedTxn = appCreateTxn.signTxn(this.deployer.sk);
      const txnResult = await this.algorandClient.client.algod.sendRawTransaction(signedTxn).do();
      
      console.log(`Transaction sent: ${txnResult.txId}`);

      // Wait for confirmation
      const confirmedTxn = await algosdk.waitForConfirmation(
        this.algorandClient.client.algod,
        txnResult.txId,
        4
      );

      const appId = confirmedTxn['application-index'];
      const appAddress = algosdk.getApplicationAddress(appId);

      console.log('✅ Contract deployed successfully!');
      console.log(`App ID: ${appId}`);
      console.log(`App Address: ${appAddress}`);
      console.log(`Transaction ID: ${txnResult.txId}`);

      return {
        success: true,
        appId,
        appAddress,
        txnId: txnResult.txId,
      };

    } catch (error) {
      console.error('❌ Deployment failed:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Get compiled approval program
   * This would normally be compiled from the TypeScript contract
   */
  private getApprovalProgram(): Uint8Array {
    // For now, return a placeholder
    // In a real implementation, this would compile the TypeScript contract to TEAL
    console.log('⚠️  Note: This is a placeholder. In production, compile the TypeScript contract to TEAL.');
    return new Uint8Array(0);
  }

  /**
   * Get compiled clear program
   */
  private getClearProgram(): Uint8Array {
    // For now, return a placeholder
    // In a real implementation, this would compile the TypeScript contract to TEAL
    console.log('⚠️  Note: This is a placeholder. In production, compile the TypeScript contract to TEAL.');
    return new Uint8Array(0);
  }

  /**
   * Initialize the deployed contract
   */
  async initializeContract(appId: number, entryFee: number, maxParticipants: number): Promise<DeployResult> {
    try {
      console.log(`🔧 Initializing contract ${appId}...`);

      const client = new HabitTrackerChallenge(appId, this.algorandClient.client.algod);
      
      const txn = await client.createChallenge(
        algosdk.algoToMicroAlgos(entryFee),
        Math.floor(Date.now() / 1000) + 3600, // Start in 1 hour
        maxParticipants,
        'Sample Challenge',
        'A sample habit tracker challenge',
        { sender: this.deployer.addr }
      );

      const signedTxn = txn.signTxn(this.deployer.sk);
      const txnResult = await this.algorandClient.client.algod.sendRawTransaction(signedTxn).do();

      console.log('✅ Contract initialized successfully!');
      console.log(`Initialization transaction: ${txnResult.txId}`);

      return {
        success: true,
        txnId: txnResult.txId,
      };

    } catch (error) {
      console.error('❌ Contract initialization failed:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }
}

// Main deployment function
async function main() {
  const deployer = new ContractDeployer();
  
  // Deploy contract
  const deployResult = await deployer.deployContract();
  
  if (!deployResult.success) {
    console.error('Deployment failed:', deployResult.error);
    process.exit(1);
  }

  // Initialize contract
  if (deployResult.appId) {
    const initResult = await deployer.initializeContract(
      deployResult.appId,
      1, // 1 ALGO entry fee
      30  // Max 30 participants
    );

    if (!initResult.success) {
      console.error('Initialization failed:', initResult.error);
      process.exit(1);
    }
  }

  console.log('🎉 Deployment completed successfully!');
}

// Run if called directly
if (require.main === module) {
  main().catch(console.error);
}

export { ContractDeployer };
